File: src/main/scala/core/Runner.scala
package core

import core.exception.CustomException

import scala.io.StdIn

class Runner(cooker: RiceCooker) {
  private var isRunning: Boolean = false

  def run(): Unit = {
    isRunning = true

    showMenu()

    val choice = StdIn.readInt()
    tryCatchLogger(
      () => {
        handleChoice(choice)
      },
      () => {}
    )
  }

  def stop(): Unit = {
    isRunning = false
    System.exit(0)
  }

  def showMenu(): Unit = {
    showState()
    println(
      """
        |Cooker Menu:
        |1. Plug a rice cooker in
        |2. Unplug a rice cooker
        |3. Increase rice amount by 1 cup
        |4. Increase water level by 1L
        |5. Decrease rice amount by 1 cup
        |6. Decrease water level by 1L
        |7. Cook rice (needs water)
        |8. Boil water (does not need rice)
        |0. Exit
        |Enter your choice:""".stripMargin
    )
  }

  def showState(): Unit = {
    logInfo(
      s"""
         |Cooker:
         |state: ${cooker.getStatus.powerStatus}, ${cooker.getStatus.workStatus}
         |rice level: ${cooker.getRiceAmount}
         |water level: ${cooker.getWaterLevel}
         |available space: ${cooker.computeAvailableSpace}
         |""".stripMargin
    )
  }

  def handleChoice(choice: Int): Unit = {
    choice match {
      case 1 => finallyContinuedTryCatchLogger(() => cooker.plug())
      case 2 => finallyContinuedTryCatchLogger(() => cooker.unplug())
      case 3 => finallyContinuedTryCatchLogger(() => cooker.addRice())
      case 4 => finallyContinuedTryCatchLogger(() => cooker.addWater())
      case 5 => finallyContinuedTryCatchLogger(() => cooker.decreaseRiceAmount())
      case 6 => finallyContinuedTryCatchLogger(() => cooker.decreaseWaterLevel())
      case 7 => finallyContinuedTryCatchLogger(() => cooker.cook())
      case 8 => finallyContinuedTryCatchLogger(() => cooker.boilWater())
      case 0 => stop()
      case _ => logError("Unknown choice, try again")
    }
    continue()
  }

  def continue(): Unit = {
    if (isRunning) {
      run()
    } else {
      logInfo("Rice cooker application closed")
      isRunning = false
      System.exit(0)
    }
  }

  private def finallyContinuedTryCatchLogger(tryCb: () => Unit): Unit = {
    tryCatchLogger(
      tryCb,
      () => continue()
    )
  }

  private def tryCatchLogger(tryCb: () => Unit, finallyCb: () => Unit): Unit = {
    try {
      tryCb()
    } catch {
      case customException: CustomException =>
        logError(customException.getMessage)
      case _: Throwable =>
        logError("An unexpected error occurred")
    } finally {
      finallyCb()
    }
  }

  private def logInfo(message: Any): Unit = {
    System.out.println(s"Info: $message")
  }

  private def logError(message: Any): Unit = {
    System.err.println(s"Error: $message")
  }
}

File: core/runner.rb
# frozen_string_literal: true

require_relative '../utils/logger_utils'
require_relative 'rice_cooker'

# Runner is an app runner defining which classes to run, showing the choices, it is like an interactive cli
class Runner
  attr_reader :cooker

  def initialize(cooker)
    @cooker = cooker
    @is_running = false
  end

  def run
    @is_running = true

    show_menu

    print('')
    answer = $stdin.gets.chomp
    handle_input(answer)
  end

  def stop
    @is_running = false
    Process.exit(0)
  end

  MENU_MESSAGE = " Cooker Menu:
          1. plug a rice cooker in
          2. unplug a rice cooker
          3. increase rice amount by 1 cup
          4. increase water level by 1L
          5. decrease rice amount by 1 cup
          6. decrease water level by 1L
          7. cook rice (needs water)
          8. boil water (does not need rice)
          0. Exit
          Enter your choice:"

  def show_menu
    show_state
    LoggerUtils.log_info <<~MENU
      #{MENU_MESSAGE}
    MENU
  end

  def show_state
    LoggerUtils.log_info <<~STATE
      Cooker:
        state: #{@cooker.status[:power_status]}, #{@cooker.status[:work_status]}
        rice level: #{@cooker.rice_amount},
        water level: #{@cooker.water_level},
        available space: #{@cooker.compute_available_space}
    STATE
  end

  def handle_input(answer)
    LoggerUtils.try_catch_logger(
      proc do
        number = Integer(answer)
        raise CustomError, 'unknown Argument, try again' if number.nil? || number.to_s != answer

        handle_choice(number)
      end,
      proc {}
    )
  end

  CHOICE_ACTIONS = {
    1 => :plug,
    2 => :unplug,
    3 => :add_rice,
    4 => :add_water,
    5 => :decrease_rice_amount,
    6 => :decrease_water_level,
    7 => :cook,
    8 => :boil_water,
    0 => :stop
  }.freeze

  def handle_choice(choice)
    action = CHOICE_ACTIONS[choice]
    raise ArgumentError, 'Invalid choice' unless action

    finally_continued_try_catch_logger { send(action) }
  end

  def plug
    @cooker.plug
  end

  def unplug
    @cooker.unplug
  end

  def add_rice
    @cooker.add_rice
  end

  def add_water
    @cooker.add_water
  end

  def decrease_rice_amount
    @cooker.decrease_rice_amount
  end

  def decrease_water_level
    @cooker.decrease_water_level
  end

  def cook
    @cooker.cook
  end

  def boil_water
    @cooker.boil_water
  end
end

def continue
  if @is_running
    run
  else
    LoggerUtils.log_info('Application closed')
    @is_running = false
    Process.exit(0)
  end
end

def finally_continued_try_catch_logger(&try_cb)
  LoggerUtils.try_catch_logger(try_cb, method(:continue))
end
